File: C/Archivos/ARCH_TIPOS_1.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

struct Ventas {
   int id;
   char *tipo; //C: Contado, T: Cuenta Corriente
   int fecha; //aaaammdd
   char *cliente;
   int vendedor; //1..30
   float monto;
};

int main () {
   FILE *arch;
   struct Ventas reg;
   int regs;

   arch = fopen ("datos.dat", "w");

   if (arch == NULL) {
      fprintf(stderr, "\nError al abrir el archivo\n");
      exit (1);
   }

   reg.id = 1;
   reg.tipo = "C"; //C: Contado, T: Cuenta Corriente
   reg.fecha = 20200801; //aaaammdd
   reg.cliente = "Don Carlos";
   reg.vendedor = 1; //1..30
   reg.monto = 5000.50;

   regs = fwrite (&reg, sizeof(struct Ventas), 1, arch);

   reg.id = 2;
   reg.tipo = "T"; //C: Contado, T: Cuenta Corriente
   reg.fecha = 20200801; //aaaammdd
   reg.cliente = "Anibal Lecter";
   reg.vendedor = 3; //1..30
   reg.monto = 4500.50;

   regs += fwrite (&reg, sizeof(struct Ventas), 1, arch);

   if(regs == 2)
      printf("Se han guardado correctamente los datos!\n");
   else
      printf("Error al guardar!\n");

   fclose (arch);

   //abro para leer
   //FILE *inf;

   arch = fopen ("datos.dat", "r");
   if (arch == NULL) {
      fprintf(stderr, "\nError al abrir para leer el archivo...\n");
      exit (1);
   }

   while(fread(&reg, sizeof(struct Ventas), 1, arch))
      printf ("Id = %d Fecha = %d Nombre = %s Monto = %f\n", reg.id, reg.fecha, reg.cliente, reg.monto);

   fclose (arch);

   return 0;
}

File: Python/Archivos/ARCH_TIPOS_1.py
import sys
import struct

class Ventas:
    def __init__(self, id_, tipo, fecha, cliente, vendedor, monto):
        self.id = id_
        self.tipo = tipo
        self.fecha = fecha
        self.cliente = cliente
        self.vendedor = vendedor
        self.monto = monto

def main():
    # escribir
    try:
        with open("datos.dat", "wb") as arch:
            reg1 = Ventas(1, "C", 20200801, "Don Carlos", 1, 5000.50)
            reg2 = Ventas(2, "T", 20200801, "Anibal Lecter", 3, 4500.50)
            # pack each record: id i, tipo char, fecha i, cliente string (we'll use fixed 50?), vendedor i, monto f
            # For simplicity, we'll write as text? but original C writes binary struct with pointers (dangerous).
            # We'll mimic by writing fixed-length fields: tipo 1 char, cliente 50 chars.
            data1 = struct.pack('i c i 50s i f', reg1.id, reg1.tipo.encode(), reg1.fecha,
                                reg1.cliente.encode().ljust(50)[:50], reg1.vendedor, reg1.monto)
            data2 = struct.pack('i c i 50s i f', reg2.id, reg2.tipo.encode(), reg2.fecha,
                                reg2.cliente.encode().ljust(50)[:50], reg2.vendedor, reg2.monto)
            arch.write(data1)
            arch.write(data2)
    except IOError as e:
        print(f"Error al abrir el archivo: {e}")
        sys.exit(1)
    print("Se han guardado correctamente los datos!")
    # leer
    try:
        with open("datos.dat", "rb") as arch:
            while True:
                data = arch.read(struct.calcsize('i c i 50s i f'))
                if not data:
                    break
                id_, tipo_bytes, fecha, cliente_bytes, vendedor, monto = struct.unpack('i c i 50s i f', data)
                tipo = tipo_bytes.decode().rstrip('\x00')
                cliente = cliente_bytes.decode().rstrip('\x00')
                print(f"Id = {id_} Fecha = {fecha} Nombre = {cliente} Monto = {monto}")
    except IOError as e:
        print(f"Error al abrir para leer el archivo: {e}")
        sys.exit(1)

if __name__ == "__main__":
    main()
